pkg/docker/config/config.go

```diff
@@ -0,0 +1,271 @@
+	}
+	encodedCreds := encodeUserAndPass(username, password)
+	return addAuthToJSON(registry, encodedCreds, path)
+}
+
+// GetAuthentication returns the value of the registry url stored in
+// the auth.json file
+func GetAuthentication(ctx *types.SystemContext, registry string) (string, string, error) {
+	if ctx != nil && ctx.DockerAuthConfig != nil {
+		return ctx.DockerAuthConfig.Username, ctx.DockerAuthConfig.Password, nil
+	}
+
+	// Check containers-storage auth.json file first
+	path, err := getPathToAuth()
+	if err != nil {
+		return "", "", errors.Wrapf(err, "error getting path to auth file")
+	}
+	auth, err := findAuthentication(registry, path)
+	if err != nil {
+		return "", "", err
+	}
+	if auth != "" {
+		return decodeDockerAuth(auth)
+	}
+
+	// check .docker/config.json as secondary option
+	path = filepath.Join(homedir.Get(), dockerCfg, dockerCfgFileName)
+	auth, err = findAuthentication(registry, path)
+	if err != nil {
+		return "", "", err
+	}
+	if auth != "" {
+		decodeDockerAuth(auth)
+	}
+
+	return "", "", nil
+}
+
+// RemoveAuthentication deletes the credentials stored in auth.json
+func RemoveAuthentication(registry string) (string, error) {
+	path, err := getPathToAuth()
+	if err != nil {
+		return "", errors.Wrapf(err, "error getting path to auth file")
+	}
+
+	auths, err := readJSONFile(path)
+	if err != nil {
+		return "", errors.Wrapf(err, "error reading JSON file %q", path)
+	}
+	if _, ok := auths.AuthConfigs[registry]; ok {
+		delete(auths.AuthConfigs, registry)
+	} else if _, ok := auths.AuthConfigs[normalizeRegistry(registry)]; ok {
+		delete(auths.AuthConfigs, normalizeRegistry(registry))
+	} else {
+		return "", ErrNotLoggedIn
+	}
+
+	file, err := os.OpenFile(path, os.O_RDWR, 0755)
+	if err != nil {
+		return "", errors.Wrapf(err, "error opening file %q", path)
+	}
+	defer file.Close()
+
+	newData, err := json.MarshalIndent(auths, "", "\t")
+	if err != nil {
+		return "", errors.Wrapf(err, "error marshaling JSON %q", path)
+	}
+
+	if err = ioutil.WriteFile(path, newData, 0755); err != nil {
+		return "", errors.Wrapf(err, "error writing to file %q", path)
+	}
+	return registry, nil
+}
+
+// getPath gets the path of the auth.json file since the user can set their own path
+// by setting CRIO_AUTH_PATH
+// If CRIO_AUTH_PATH is not set, the default path is ~/.config/containers/auth.json
+func getPathToAuth() (string, error) {
+	if os.Getenv("CRIO_AUTH_PATH") != "" {
+		return filepath.Join(os.Getenv("CRIO_AUTH_PATH"), authCfgFileName), nil
+	}
+	dir := filepath.Join(homedir.Get(), authCfg)
+	if _, err := os.Stat(dir); os.IsNotExist(err) {
+		if err = os.Mkdir(dir, 0600); err != nil {
+			return "", errors.Wrapf(err, "error creating directory %q", dir)
+		}
+	}
+	return filepath.Join(dir, authCfgFileName), nil
+}
+
+// readJSONFile unmarshals the authentications stored in the auth.json file
+func readJSONFile(path string) (dockerConfigFile, error) {
+	file, err := os.OpenFile(path, os.O_RDWR|os.O_CREATE, 0755)
```

Comment:
Why is a ”read” function modifying the filesystem at all?

AFAICT this could be more or less`ioutil.ReadFile`+`os.IsNotExist`+`json.Unmarshal`.

Reply:
It is possible for the file to not exist, so this creates the file if it doesn't otherwise just reads from it. Thought this was the best place to check for existence as its the first place the file is trying to be read.

Reply:
It seems far clearer to me to create the file only when trying to update it and there is no previous version.

It’s unexpected that a reader modifies the filesystem; if somebody wanted to do very strict hardening with SELinux or the like, a process intended to read the file might not have write permissions.

And we don’t actually want “the file” to always exist, because `GetAuthentication` will call `readJSONFile` for all of the paths (`~/.config/containers/auth.json`, `~/.docker/config.json`, `~/.dockercfg`) in turn if none of them contain credentials for the hostname in question.  So we would end up creating the two legacy files as well.  

Finally, creating the file first forces the code to do the “size != 0” checks, which is an extra hassle, and imprecise, because zero-sized file can happen in other ways than by creating a new file. For example if something, e.g. a shutdown in the middle of an update, corrupts the filesystem and truncates a previously-existing config, the user would probably benefit from an error message saying that a zero-bytes file is invalid JSON).

Reply:
so I should create the file in `addAuthToJSON`? since thats where an update to the file happens?

Reply:
Yes; actually `addAuthToJSON` is already creating the file — twice. Just the `ioutil.WriteFile` should be enough.